static/src/javascripts/projects/commercial/modules/messenger.js

```diff
@@ -0,0 +1,240 @@
+        // Because any listener can have side-effects (by unregistering itself),
+        // we run the promise chain on a copy of the `LISTENERS` array.
+        // Hat tip @piuccio
+        const promise = LISTENERS[data.type]
+            .slice()
+            // We offer, but don't impose, the possibility that a listener returns
+            // a value that must be sent back to the calling frame. To do this,
+            // we pass the cumulated returned value as a second argument to each
+            // listener. Notice we don't try some clever way to compose the result
+            // value ourselves, this would only make the solution more complex.
+            // That means a listener can ignore the cumulated return value and
+            // return something else entirely—life is unfair.
+            // We don't know what each callack will be made of, we don't want to.
+            // And so we wrap each call in a promise chain, in case one drops the
+            // occasional fastdom bomb in the middle.
+            .reduce(
+                (func, listener) =>
+                    func.then(ret => {
+                        const thisRet = listener(
+                            data.value,
+                            ret,
+                            getIframe(data)
+                        );
+                        return thisRet === undefined ? ret : thisRet;
+                    }),
+                Promise.resolve(true)
+            );
+
+        return promise
+            .then(response => {
+                respond(null, response);
+            })
+            .catch(ex => {
+                reportError(ex, {
+                    feature: 'native-ads',
+                });
+                respond(formatError(error500, ex), null);
+            });
+    } else if (typeof LISTENERS[data.type] === 'function') {
+        // We found a persistent listener, to which we just delegate
+        // responsibility to write something. Anything. Really.
+        LISTENERS[data.type](respond, data.value, getIframe(data));
+    } else {
+        // If there is no routine attached to this event type, we just answer
+        // with an error code
+        respond(formatError(error405, data.type), null);
+    }
+};
+
+const on = (window: Object): void => {
+    window.addEventListener('message', onMessage);
+};
+
+const off = (window: Object): void => {
+    window.removeEventListener('message', onMessage);
+};
+
+export const register = (
+    type: string,
+    callback: () => void,
+    options: Object = {}
+): void => {
+    if (REGISTERED_LISTENERS === 0) {
+        on(options.window || window);
+    }
+
+    /* Persistent LISTENERS are exclusive */
+    if (options.persist) {
+        LISTENERS[type] = callback;
+        REGISTERED_LISTENERS += 1;
+    } else {
+        // set LISTENERS[type] to an empty array, if it is currently undefined or null
+        if (!LISTENERS[type]) {
+            LISTENERS[type] = [];
```

Comment:
Yeah I'm not surprised. This should work though: `listeners[type] = liteners[type] || []`?

Minor bug in Flow's data flow analysis